### PR TITLE
Add a subset of OpenGL shaders from Tyrells repo

### PR DESCRIPTION
**Pull request by johnnovak**
This adds a subset of the SVN shaders from Tyrells' repo: https://github.com/tyrells/dosbox-svn-shaders

Tyrells seems to be uninterested in maintaining these shaders, therefore we are incorporating them into the project and taking over their maintenance. I've applied some minor updates that were made to them since Tyrells last touched these files.

I've also converted a new one called `zfast_crt_composite` for fun and to see how hard it is. Turns out it's quite easy, so I'll convert a few more of the new additions later.

I've ommitted some shaders from his repo for the following reasons:

- These `crt` shaders appear to be obviously broken:
   - `ScanLine.glsl`
   - `crt-cgwg-fast.glsl`
   - `crt-hyllian-3d.glsl`
   - `smuberstep-glow.glsl`
- The `interpolation` shaders implement pixel-perfect scaling but with a hardcoded 1.2:1 aspect ratio stretch which will yield wrong results for us, plus we already have our own pixel-perfect mode anyway. `pixellate` doesn't seem too useful either as we already have our superior `sharp` shader. Overall, they would just needlessly confuse users while not bringing anything new or useful to the table.
- My executive summary of the `xbr` shaders is that they look like crap; I can't imagine anyone wanting to use these...

Overall, I don't want to include obviously broken and/or bad looking shaders in our official releases because that would make us look like amateurs.

I've also added helpful hints if the old shader names are detected so users can update their configs more easily.

### CODE REVIEW

**Comment by GranMinigun** [comment]
Throwing this in just so I don't forget about it forever: [catmull-rom.glsl.txt](https://github.com/dosbox-staging/dosbox-staging/files/10251700/catmull-rom.glsl.txt)

**Comment by johnnovak** [comment]
Btw, it is possible that the below excluded CRT shaders only need a few compatibility tweaks, but they're definitely broken on my shiny new M1 Mac. So maybe we can revisit them later; right now I don't have the capacity to investigate them in great detail.

- ScanLine.glsl
- crt-cgwg-fast.glsl
- crt-hyllian-3d.glsl
- smuberstep-glow.glsl

**Comment by kcgen** [comment]
In the first commit, I think that's the the "bulk add" from Tyrell's repo; suggest setting Tyrells as that commit author, then with yor further additions and fixes in follow on commit(s).

File: src/gui/render.cpp [hunk 1]
```diff
@@ -633,6 +634,43 @@ bool RENDER_UseSRGBFramebuffer()
 
 #endif
 
+#if C_OPENGL
+void log_warning_if_legacy_shader_name(const std::string name)
```
Comment (kcgen) [inline src/gui/render.cpp:638]: Suggest a reference (Avoid re-copying content)

File: src/gui/render.cpp [hunk 2]
```diff
@@ -633,6 +634,43 @@ bool RENDER_UseSRGBFramebuffer()
...
+	        {"crt-fakelottes-flat", "crt/fakelottes"},
+	        {"rgb2x", "scaler/rgb2x"},
+	        {"rgb3x", "scaler/rgb3x"},
+	        {"scan2x", "scaler/scan2x"},
+	        {"scan3x", "scaler/scan3x"},
+	        {"sharp", "interpolation/sharp"},
+	        {"tv2x", "scaler/tv2x"},
+	        {"tv3x", "scaler/tv3x"}};
+
+	std_fs::path path = name;
```
Comment (kcgen) [inline src/gui/render.cpp:655]: "path" feels a bit too much like a type itself (Like: `std::set set = {}`)

File: src/gui/render.cpp [hunk 3]
```diff
@@ -633,6 +634,43 @@ bool RENDER_UseSRGBFramebuffer()
...
+	std_fs::path ext  = path.extension();
+
+	if (!(ext == "" || ext == ".glsl")) {
+		return;
+	}
```
Comment (kcgen) [inline src/gui/render.cpp:660]: If someone uses `glshader = tv2x.glsl`, then we don't warn (`ext == ".glsl"`)?

File: src/gui/render.cpp [hunk 4]
```diff
@@ -633,6 +634,43 @@ bool RENDER_UseSRGBFramebuffer()
...
+
+	path.replace_extension("");
+
+	auto it = legacy_name_mappings.find(path.string());
```
Comment (kcgen) [inline src/gui/render.cpp:664]: const

File: src/gui/render.cpp [hunk 5]
```diff
@@ -633,6 +634,43 @@ bool RENDER_UseSRGBFramebuffer()
...
+	if (it != legacy_name_mappings.end()) {
+		auto new_name = it->second;
```
Comment (kcgen) [inline src/gui/render.cpp:666]: const

File: src/gui/render.cpp (hunk 1 above)
Comment (johnnovak) [reply src/gui/render.cpp:638]: Done.

File: src/gui/render.cpp (hunk 2 above)
Comment (johnnovak) [reply src/gui/render.cpp:655]: Changed.

File: src/gui/render.cpp (hunk 3 above)
Comment (johnnovak) [reply src/gui/render.cpp:660]: We do. We just exclude names here that do _not_ end with `.glsl` or are extensionless.

Then if the extension was `.glsl`, we remove it with `shader_path.replace_extension("");`.

This way the size of `legacy_name_mappings` could be cut in half (i.e. we don't need to list both names, once with and once without the `.glsl` extension). I just liked it better this way, but can include both versions in the map and remove this logic if you think it's too confusing.

File: src/gui/render.cpp (hunk 4 above)
Comment (johnnovak) [reply src/gui/render.cpp:664]: Done.

File: src/gui/render.cpp (hunk 5 above)
Comment (johnnovak) [reply src/gui/render.cpp:666]: Done.

**Comment by johnnovak** [comment]
> In the first commit, I think that's the the "bulk add" from Tyrell's repo; suggest setting Tyrells as that commit author, then with yor further additions and fixes in follow on commit(s).

Ok, sounds fair. I didn't do it because he isn't the original author of these shaders, but then, neither am I 😉

**Comment by kcgen** [comment]
> > In the first commit, I think that's the the "bulk add" from Tyrell's repo; suggest setting Tyrells as that commit author, then with yor further additions and fixes in follow on commit(s).
> 
> Ok, sounds fair. I didn't do it because he isn't the original author of these shaders, but then, neither am I wink

If it's possible to "right some wrongs", and someone else put together some of those GLSL files as-is and you can put their name against them (then please so!).

But if the shaders existed in other forms laguage construct (say, Retroarch), and Tyrell was the person who "ported" it to DOSBox-compatible GLSL, then yeah - he's the closest connection to the derived GLSL file (and I see the original authors are in the text itself.. which is perfect).

Really, it's just about connecting some dots for future readers, if it's easy enough to track down! :+1: 

Oh - I had forgot where this reference was, but here's what we use for all upstream commits:

https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages-for-patches-authored-by-someone-else

**Comment by johnnovak** [comment]
> But if the shaders existed in other forms (say, Retroarch), and Tyrells was the person who ported the algorithm to DOSBox-compatible GLSL format, then yeah - he's probably the right person to mark as the actual file author (as a derived work).

Yeah, technically the porting boils down to renaming a few vars, and rearranging the code a little bit here and there. Plus introducing some handling for our "no-bilinear" modes. Otherwise, the original code is 1:1 there without changes, he even kept the wonky formatting of the originals 😁 

I'll just attribute him, that will do. Also, it's a common practice for people to "remix" these shaders by taking bits and pieces from each then creating their own variants; really hard to track who did what (but I can recognise certain copy-pasted code fragments in quite a few of them 😄 )

**Comment by johnnovak** [comment]
Well, the actual chain of event for many are

- somebody wrote the first version
- someone ported them to RetroArch, but usually not in GLSL format
- someone backported a bunch of RetroArch shaders into GLSL
- Tyrells ported these GLSL backports of ports to DOSBox

You get the picture... 😉

**Comment by kcgen** [comment]
Sounds like a very fluid family tree :sweat_smile: 
Thanks for the background; sounds like Tyrells is the closest author for the inbound GLSLs! That's enough to connect the chain, I'd say!

**Comment by johnnovak** [comment]
> Sounds like a very fluid family tree 😅 Thanks for the background; sounds like Tyrells is the closest author for the inbound GLSLs. That's enough to connect the chain, I'd say!

I broke the commit apart and tried to include some information about this situation in the commit messages. Please have a look and let me what you think. If you're happy, this can be merged.

File: src/gui/render.cpp (hunk 3 above)
Comment (kcgen) [reply src/gui/render.cpp:660]: Thanks ; I misread the outer "not" applied to both `or` statements. Makes sense!

**Review by kcgen — APPROVED** [review]
If you can add @GranMinigun 's implementation of the Catmull-Rom scaler: https://github.com/dosbox-staging/dosbox-staging/files/10251700/catmull-rom.glsl.txt , that would be excellent.

Alternatively, it can be added after this PR too.

**Comment by kcgen** [comment]
> [Add interpolation/catmull-rom shader](https://github.com/dosbox-staging/dosbox-staging/pull/2137/commits/e22e996aef963ee9b3348948835fd2a62fe7b41a)

:+1:

**Comment by johnnovak** [comment]
> If you can add @GranMinigun 's implementation of the Catmull-Rom scaler: https://github.com/dosbox-staging/dosbox-staging/files/10251700/catmull-rom.glsl.txt , that would be excellent.
> 
> Alternatively, it can be added after this PR too.

Done.

Now we have a crap LCD-style interpolator as well for those poor people who never experienced a CRT :trollface: 

@GranMinigun I think the shader is gamma-incorrect; note the dark band around the guy's tie as it blends into the cyan shirt. That's the opening scene of [Zak McKracken and the Alien Mindbenders](https://en.wikipedia.org/wiki/Zak_McKracken_and_the_Alien_Mindbenders), by the way.

image

**Comment by GranMinigun** [comment]
> I'm assuming you did the port to the .GLSL file?

Correct.

> It's definitely less crap (=sharper) than bilinear, though, I'll give it that.

It's not intended for low-res games, I want it for SVGA games that don't look right to my eye with sharp bilinear shader. Those that feature a lot of gradients or pre-rendered graphics, for example. Performance penalty is quite high on my machine, by the way - around 3 ms.

> I think the shader is gamma-incorrect; note the dark band around the guy's tie as it blends into the cyan shirt.

I tried applying sRGB pragmas, but they didn't change a thing. Also note that bicubic interpolation is prone to overshooting in high-contrast scenarios (Catmull-Rom is just a name of a bicubic spline). Feel free to correct sampling if there are indeed problems with gamma.

**Comment by JohnnyWalkerDigital** [comment]
What happened here? Many of the shaders have disappeared? eg. crt/geom.tweaked and crt/geom

**Comment by johnnovak** [comment]
> What happened here? Many of the shaders have disappeared? eg. crt/geom.tweaked and crt/geom

What? Why are you commenting on a 2022 PR?

**Pull request merged**